pinot-common/src/main/java/org/apache/pinot/common/protocols/SegmentCompletionProtocol.java

```diff
@@ -180,6 +184,16 @@ private Request(Params params, String msgType) {
     }
 
     public String getUrl(String hostPort, String protocol) {
+      String streamPartitionMsgOffset;
+      try {
+        streamPartitionMsgOffset = _params.getStreamPartitionMsgOffset() == null ? null :
+            URLEncoder.encode(_params.getStreamPartitionMsgOffset(), StandardCharsets.UTF_8.toString());
```

Comment:
If Stream partition offset needs to be url encoded then that logic has to be specific to the stream partition offset of the stream implementation. Please move it there. As far as the protocol is concerned, it should be a string.

Protocol getters should not be throwing exceptions

Reply:
That would mean making the `toString` method of the `StreamPartitionOffset` impl return a url encoded string, which is incorrect. The impl shouldn't always url encode the toString representation. I see 2 options:
1. Adding a getURLEncodedString method in the impl
2. Move this logic inside `withStreamPartitionMsgOffset`. I think it is fair to make this method assume that stream partition offsets can be complex string representations and that it should encode them before accepting them into params.
Lmk which option you prefer.

Reply:
I think I see where you are headed, let us discuss in person and iron out this (and any other)

Reply:
How do we deal with the segment name parameter? What happens if we assign some string value as a segment name that will be broken on the receiver side without going through the URL encoding?

IMO, we should encode to URL format for everything. Whatever value provided to the `_params`, the same value should be read by the component that will parse this request.

Reply:
It makes sense to urlencode, I agree. Check with PRs 4426 and 5639 I think we had a case before where we utl encoded segment name, and it caused some issues

Reply:
@snleee made the change to encode everything. Also used a URIBuilder instead of string concatenation.
Checked the 2 PRs. I have added tests to make sure the segment name encodes correctly.